### PR TITLE
Add Mission Planner manifest

### PR DESCRIPTION
**Pull request by G0rocks**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ x] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?
- [ x] Have you validated your manifest locally with `winget validate <manifest>`, where `<manifest>` is the name of the manifest you're submitting?
- [x] Have you tested your manifest locally with `winget install -m <manifest>`?

-----
I had problems testing the manifest locally, could somebody help me out?
I got this error message:
`winget-pkgs\manifests\Ardupilot\MissionPlanner>winget install -m 1.3.72.yaml
Found Mission Planner [Mission.Planner]
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licences to, third-party packages.
Downloading https://firmware.ardupilot.org/Tools/MissionPlanner/MissionPlanner-latest.msi
An unexpected error occurred while executing the command:
InternetOpenUrl() failed.
0x80072f19 : unknown error`

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-pkgs/pull/8521)

### CODE REVIEW

**Comment by G0rocks** [comment]
I have now successfully cleared all tests required

**Comment by G0rocks** [comment]
Aaaaawww yeah, all checks passed :)

**Comment by G0rocks** [comment]
I did not add those switches nor forget to do so. I attempted to mostly follow the minimal manifest syntax listed here: https://github.com/microsoft/winget-pkgs, though I did add something I found in other YAML files which looked cool.
I do not know a lot about YAML and manifests in general, I just thought this idea of a windows package manager is awesome. I'll look into this thing this time but you are putting constraints on the "create manifest" project afterwards. If these silent or silentwithprogress switches are required, please add that information to the main page of this repository so people know exactly what they are getting into when they decide to make a manifest file. I'll make an issue :) (The issue is here: https://github.com/microsoft/winget-pkgs/issues/8543)

But I'm having a hard time working out these silent switches, could you help me out? How do I add them and what is the syntax in the YAML file? Thanks.

**Comment by yigitemres** [comment]
Hi there,

Thanks for your contribution. I think that there is something wrong with msfbot right now, because there is no need to adding `Switches` to MSI manifests. All the MSI uses the same switches (if the installer doesn't need any additional arguments) and because of this winget itself uses MSI switches internally.

> But I'm having a hard time working out these silent switches, could you help me out?

```powershell
~/Desktop
❯ .\MissionPlanner-latest.msi /help
```

You can use /help, /?, -?, -help, --help, -h and --h. When command works successfully you see this window.

msiexec_BBy5EecxDF

`/passive` is correspond to `SilentWithProgress` switch and `/quiet` is correspond to `Silent` switch.

```yaml
Id: Ardupilot.MissionPlanner
Version: 1.3.72
Name: Mission Planner
Publisher: ArduPilot development team
License: GNU General Public License v3.0
Description: Mission Planner is a full-featured ground station application for the ArduPilot open source autopilot project.
InstallerType: msi
Installers:
  - Arch: x64
    Url: https://firmware.ardupilot.org/Tools/MissionPlanner/MissionPlanner-latest.msi
    Sha256: 1eb59aef23977a0d6b1a493160667649d009160755e90a27843af3e6f140b75a
    Switches:
      Silent: /quiet
      SilentWithProgress: /passive
```

If you have any question feel free to ask. Have a nice day. 🥂

**Comment by G0rocks** [comment]
Aha, okay. I understand a bit better. Thanks :)

Should I close the created issue? #8543 

Does this mean that everything is ready and we can merge the request or is there something else I need to do?

Thanks 🥂 You have a nice day too ^_^

**Comment by yigitemres** [comment]
Issue can be stayed because we need step by step guide for new contributors (IMHO). If you can modify your issue and propose the guide, it will be better I think.

> Does this mean that everything is ready

When checks are successful, only think to do is wait for merge. Assignees are generally fast and merge immediately but sometimes there can be problems. Let say that your manifest contains download link that which build daily. When you submit your pr initial check are passed. But when assignee try to merge it, bot checks again and if there is new build, it causes hash error. When this happens you have to find new hash and push new commit. And bot checks again. This problem only exist on new manifests. If manifest exist on the repo, wingetbot automates hash checks and creates pr with new hash.

**Comment by KevinLaMS** [comment]
Hey @G0rocks  
Sorry about the challenges.
I will look at improving the error message for the bot so as not to lead you down the wrong path.
@denelon and I will ensure we get a 'guide' to improve the process.  

kevin

**Comment by KevinLaMS** [comment]
@G0rocks So it appears to block on a dialog.
image

it appears to be a bunch of drivers.
image

Any idea how to not block install?

**Comment by G0rocks** [comment]
> Hey @G0rocks
> Sorry about the challenges.
> I will look at improving the error message for the bot so as not to lead you down the wrong path.
> @denelon and I will ensure we get a 'guide' to improve the process.
> 
> kevin

Thanks Kevin, I appreciate it.

> Any idea how to not block install?

So I didn't have space for the files on my C:drive so I ran this install with the -i flag and then for the Laser Navigation Ports (COM & LPT) I clicked "Don't install" and the installation totally works for me. Other than that I'm sorry but I don't totally know how this package manager works. Maybe we can specify in the YAML file that winget should say "Don't install" to those extra drivers? Or something similar? Or maybe an issue should be created where if winget is installing something and gets stuck it will show it to the user and ask them for help (unless that already exists)? But that doesn't mean we shouldn't fix this stuff before adding a package.

**Comment by yigitemres** [comment]
If we can't modify `CustomAction` without `Property` from shell, we can't do anything about it. Somehow we have to modify `Install_signed_Driver64` action `[Drivers]DPInstx64.exe` to `[Drivers]DPInstx64.exe /S`

**Comment by G0rocks** [comment]
Well that's a little over my head, what is install_signed_driver64?

**Comment by yigitemres** [comment]
@G0rocks Every MSI contains tables. These tables contains various infos like hashes, which file exist in the installer etc. Also these tables contains information about installation sequences. If you open MSI with [Orca](https://docs.microsoft.com/en-us/windows/win32/msi/orca-exe) you can see these tables.

Orca_flqe9Zj0mv

If you remember when you use `.\MissionPlanner-latest.msi /help` this command, at the bottom of the text there is `Property` part (`Setting Public Properties - [PROPERTY=PropertyValue]`). You can change public properties by shell. (`.\MissionPlanner-latest.msi /silent Manufacturer="Stephen King"`) Some devs includes `Property` in order to modify `CustomAction`. Generally when using `/quiet` or `/passive` command some `CustomActions` are gonna be ignored (it's about how the devs structure their installer) and because of this there is no need to modify any `CustomAction` from `Property`. But in this situation, `MissionPlannet-latest.msi` always gonna call `Install_signed_Driver86` or `Install_signed_Driver64` if it's not installed.

Orca_CahDImAaT5

Actually this is a good thing when you use installer directly. But in our situation this cause stalls/timed out for winget.

Orca_yeZMI3ZlSQ

As you can see there is two `CustomAction` named `Install_signed_Driver86` and `Install_signed_Driver64`. If we can somehow modify these parts and add silent/quiet switches (like `[Drivers]DPInstx64.exe /S`) we have no problem. But we can't directly modify `CustomAction` (as far as I know) and we need `Property` to modify them (and this installer don't have any `Property` for modifying these `CustomActions`)

Long story short, either this installer gonna be force pushed with existing time out (this is happened for some installers like TeX Live. But maybe there is some conditions to met. Assignees gonna decide) or someone told us how we can modify `CustomAction` part.

Have a nice day! 🥂

**Comment by G0rocks** [comment]
Okay okay so if I understand you correctly we need to modify the `CustomAction` in the mission planner installer so that it will accept switches and not attempt to install those drivers if the switches are silent? Or do you mean that we need to modify winget to deal with it?
Thanks :)

**Comment by yigitemres** [comment]
@G0rocks (I have very limited knowledge about how MSI works) In order to modify `CustomAction` part, we need `Property` to modify that specific part. Without `Property` we can't modify `CustomAction` (Only way to send args to installer is `Property`). We don't try to prevent installing drivers. Adding /S or /Q installs drivers without need for user interaction (just like how we use winget). If you look `InstallExecuteSequence` (2nd screenshot) you will see that drivers always gonna installed if they are not installed. I'm not sure adding args to `CustomAction` will change anything but it will work I think. If MSI don't work like this way, we need a new feature for winget about driver bundled installers.

**Comment by G0rocks** [comment]
Okay, so if I understand you correctly, we need to add a `property` to the mission planner MSI (which is here btw: https://github.com/ArduPilot/MissionPlanner/tree/master/Msi) which allows mission planner to be installed quietly, given the correct argument?
@KevinLaMS could you help us out with this?

**Comment by yigitemres** [comment]
@G0rocks Installer only need this; when we pass `/passive` or `/quiet` args, installer also needs pass these (driver installer user /S and /Q) arguments to driver installers.

**Comment by G0rocks** [comment]
Right, the mission planner installer, correct?

**Comment by yigitemres** [comment]
1. When we `MissionPlanner-latest.msi /passive`, driver installer should `DPInstx64.exe /S`.
2. When we `MissionPlanner-latest.msi /quiet`, driver installer should `DPInstx64.exe /Q`.

**Comment by G0rocks** [comment]
Okay but I'm wondering which program should take care of that. Winget or the mission planner installer?

**Comment by G0rocks** [comment]
I edited the format to fulfill the changed requirements for the yaml file.
Also started a discussion on the Sha256 values [here](https://github.com/microsoft/winget-cli/discussions/937)

**Comment by G0rocks** [comment]
Ok, I'm back to being confused. The hash changes everytime I run `winget hash <yaml>`. I posted a question in the discussions, [here](https://github.com/microsoft/winget-cli/discussions/937#discussioncomment-708949), I also added an issue a while back, [here](https://github.com/microsoft/winget-pkgs/issues/8546), asking for someone to add a "How to figure out the Sha256 stuff" section, for example to [this page](https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md).
Though all of that is probably pointless since this installer isn't silent (or I don't think so at least)

**Comment by denelon** [comment]
G0rocks I downloaded the installer twice in a browser signed in and once in a private session to see if the source is changing the file per download (unlikely, but possible). The most common scenario we see with different hash values is a publisher rolling out a new binary at a "vanity URL" using a content delivery network. I got the same hash value all three times:

```
PS C:\Users\denelon\Desktop> winget hash .\MissionPlanner-latest.msi
Sha256: 1eb59aef23977a0d6b1a493160667649d009160755e90a27843af3e6f140b75a
PS C:\Users\denelon\Desktop> winget hash .\MissionPlanner-latest2.msi
Sha256: 1eb59aef23977a0d6b1a493160667649d009160755e90a27843af3e6f140b75a
PS C:\Users\denelon\Desktop> winget hash .\MissionPlanner-latest3.msi
Sha256: 1eb59aef23977a0d6b1a493160667649d009160755e90a27843af3e6f140b75a
```
missionplanner

File: manifests/a/Ardupilot/MissionPlanner/1.3.72/Ardupilot.MissionPlanner.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+PackageLocale: "en-US"
+Publisher: "ArduPilot development team"
+PackageName: "Mission Planner"
+License: "GNU General Public License v3.0"
+ShortDescription: "Mission Planner is a full-featured ground station application for the ArduPilot open source autopilot project."
+Installers:
+  - Architecture: "x64"
+    InstallerType: "msi"
+    InstallerUrl: "https://firmware.ardupilot.org/Tools/MissionPlanner/MissionPlanner-latest.msi"
+    InstallerSha256: f9401a1857c447ef59769ef33414c4df4dd71ec9d90523e632a483cec60e529e
```
Comment (denelon) [inline manifests/a/Ardupilot/MissionPlanner/1.3.72/Ardupilot.MissionPlanner.yaml:12]: 1eb59aef23977a0d6b1a493160667649d009160755e90a27843af3e6f140b75a

**Comment by G0rocks** [comment]
So it says installer validation failed but it works on my end. Using the hash from denelon ¯\_(ツ)_/¯

**Comment by G0rocks** [comment]
I think working on this is pointless until [this issue](https://github.com/ArduPilot/MissionPlanner/issues/2609) is resolved, which I don't know how to do, or until winget can install with installers that don't install silently. So just going to take a break

**Comment by denelon** [comment]
@G0rocks can you rebase to see if that fixes the Pull Request error?

**Review by ItzLevvie — CHANGES_REQUESTED** [review]
image

**Comment by ItzLevvie** [comment]
cc @G0rocks

**Comment by G0rocks** [comment]
I am just going to try again from here:
https://github.com/microsoft/winget-pkgs/pull/432539
Set it up from scratch, this time with komac

Thanks :)
